### ISSUE TITLE
"Community Goal" speech responder triggered multiple times

### ISSUE THREAD

**Issue by CMDR-Cometborne** [EDCD/EDDI#123]
## Steps to reproduce
Sign up for multiple community goals at the same time, then trigger a CommunityGoal journal message

## Expected
There should be one message per community goal.

## Observed
What I am getting is the messages for all goals repeated as many times as I have active goals.  So when I have two goals, I get four messages:
1) goal 1
2) goal 2
3) goal 1
4) goal 2

With five community goals, I am getting 25 messages (goals 1,2,3,4,5, 1,2,3,4,5, 1,2,3,4,5, 1,2,3,4,5, 1,2,3,4,5).

## Investigation
Checked the journal, and only one CG line is written.  Below is the version with 5 CGs, raw from the journal file.

{ "timestamp":"2017-09-28T21:02:43Z", "event":"CommunityGoal", "CurrentGoals":[ { "CGID":420, "Title":"Setting up the Candy Shop", "SystemName":"Irusan", "MarketName":"Bella Port", "Expiry":"2017-10-05T15:00:00Z", "IsComplete":false, "CurrentTotal":502881, "PlayerContribution":0, "NumContributors":769, "TopRankSize":10, "PlayerInTopRank":false, "TierReached":"Tier 1", "PlayerPercentileBand":100, "Bonus":100000 }, { "CGID":421, "Title":"Winged Hussars Appeal", "SystemName":"HIP 110028", "MarketName":"Zander Dock", "Expiry":"2017-10-05T15:00:00Z", "IsComplete":false, "CurrentTotal":10603, "PlayerContribution":0, "NumContributors":108, "TopRankSize":10, "PlayerInTopRank":false, "TierReached":"Tier 2", "PlayerPercentileBand":100, "Bonus":200000 }, { "CGID":422, "Title":"Second Aegis Initiative", "SystemName":"Wolf 397", "MarketName":"Chelbin Service Station", "Expiry":"2017-10-05T15:00:00Z", "IsComplete":false, "CurrentTotal":540654, "PlayerContribution":0, "NumContributors":598, "TopRankSize":10, "PlayerInTopRank":false, "TierReached":"Tier 2", "PlayerPercentileBand":100, "Bonus":200000 }, { "CGID":418, "Title":"Engineers Appeal for Thargoid Material – Professor Palin", "SystemName":"Maia", "MarketName":"Obsidian Orbital", "Expiry":"2017-09-28T15:00:00Z", "IsComplete":true, "CurrentTotal":14395838, "PlayerContribution":10196, "NumContributors":7862, "TopRankSize":10, "PlayerInTopRank":false, "TierReached":"Tier 3", "PlayerPercentileBand":10, "Bonus":12790800 }, { "CGID":419, "Title":"Engineers Appeal for Thargoid Material – Liz Ryder", "SystemName":"Eurybia", "MarketName":"Awyra Flirble", "Expiry":"2017-09-28T15:00:00Z", "IsComplete":true, "CurrentTotal":18082575, "PlayerContribution":10192, "NumContributors":8723, "TopRankSize":10, "PlayerInTopRank":false, "TierReached":"Tier 3", "PlayerPercentileBand":10, "Bonus":12790800 } ] }

If you want the complete journal, drop me a note.

**Comment by richardbuckle** [comment]
Believed fixed in 2.4-b4. Could you please re-test?

**Comment by CMDR-Cometborne** [comment]
Nearly there. Updated to 2.4.0-b4 and switched back to the default personality. I am being informed once per community goal, so the truly annoying bug is fixed.  However, there looks to be a logic bug with the first/next/last text between goals.  With five CG active, I get:

The first <cg1 speech>
The last <cg2 speech>
The last <cg3 speech>
The last <cg4 speech>
The next <cg5 speech>, and <pause>

In other words, next and last seem to be swapped.

**Comment by Tkael** [comment]
Thanks for your feedback. Fixed for the next release. :-D

In the meantime, here's the updated script: 
```
{_ Set Context _}
{SetState('eddi_context_last_subject', 'community goal')}
{_ Save goal location to context if there is only one goal present _}
{if len(event.cgid) = 1: {SetState('eddi_context_goal_system_name', event.system[cur])} }
{if len(event.cgid) = 1: {SetState('eddi_context_goal_station_name', event.station[cur])} }

{F('Honorific')}, 

{if len(event.name) > 1:
    {OneOf("you have expressed interest in", "I am tracking")}
    the following community goal{if len(event.cgid) > 1:s}.
}

{Pause(250)}

{set cur to 0}
{while cur < len(event.name):

   {if cur = 0 && len(event.name) = 1:
      {_ nothing _}
   |elif cur = 0 && len(event.name) > 1:
      The first,
   |elif cur < (len(event.name) - 1):
      The next, 
   |else:
      The last,
   }

   {event.name[cur]},

   {if event.iscomplete[cur]:
      {OneOf("is complete", "has finished")} 
   |else: 
      {OneOf("is not yet complete", "is active", "is incomplete")} 
   }

   {if event.tier[cur]:
       {OneOf("and has reached", "and currently at")} {event.tier[cur]}.
       {if event.contribution > 0:
           Your expected reward is 
           {Occasionally(2, "currently")}
           {Occasionally(2, "calculated at")}
           {Humanise(event.tierreward[cur])} credits
           {Occasionally(2, "in the {event.percentileband[cur]}th percentile")}
       }
   }.

   {_ Create a comparison function _}
   {set compare(stat) to:     
       {if stat < 0.01:
            just a {OneOf("fraction", "token")} of            
       |elif stat < 0.10:
            less than a tenth of
       |elif stat < 0.25:
            less than a quarter of
       |elif stat < 0.5:
            less than half of
       |elif stat < 0.75:
            less than
       |elif stat > 0.89: 
            {OneOf("fractionally","just")}
            less than
       |elif stat > 0.95 && stat < 1.05:
            {OneOf("approximately","roughly","nearly")} 
            equal to 
       |elif stat < 1.25:
            a little more than
       |elif stat < 1.5:
            more than
       |elif stat < 2:
            significantly more than
       }
    }
    {set compare_contribution to (event.contribution[cur] / round(event.total[cur] / event.contributors[cur], 0))}

    {if compare_contribution != 0:
       {if event.toprank[cur]: 
          Congratulations! 
          You are one of {event.topranksize[cur]} top commanders in this event, having contributed 
          {compare(compare_contribution)}
          the average
       |else:
          {Occasionally(2, "You have contributed 
          {compare(compare_contribution)}
          the average")}
       }.
    }

    {if !event.iscomplete[cur]:
       You may 
       {if event.contribution[cur] > 0: 
          {OneOf("contribute more", "make additional contributions")} 
       |else: 
          {OneOf("contribute", "make a contribution")}
       }
       at {P(event.station[cur])}, in
       {OneOf("{P(event.system[cur])}", "the {P(event.system[cur])} system" )}.
    }

    {set cur to cur + 1}
    {Pause(1000)}
}
```
o7
